### PR TITLE
add ppt2ec pept2go pept2funct to API

### PR DESCRIPTION
**Pull request by pverscha**
Implementation of pept2funct, pept2go and pept2ec API-endpoints

_[Original pull request](https://github.ugent.be/unipept/unipept/pull/658) by @pverscha on Wed Oct 03 2018 at 16:08._
_Merged by @bmesuere on Tue Jan 15 2019 at 14:03._

### CODE REVIEW

**Comment by beardhatcode** [comment]
I think it might be cleaner if you used jbuilder instead of `Oj.dump`.
jBuilder used Oj internally, so the performance should be identical.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-79581) by @beardhatcode on Wed Oct 03 2018 at 17:06._

**Comment by pverscha** [comment]
@beardhatcode Thank you! I'll update the API-views accordingly.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-79583) by @pverscha on Wed Oct 03 2018 at 18:56._

File: app/controllers/api/api_controller.rb [hunk 1]
```diff
@@ -88,7 +89,15 @@ def pept2taxa
...
   def pept2lca
-    @result = {}
+    @result = pept2lca_helper
+
+    filter_input_order
+
+    respond_with(@result)
+  end
+
+  def pept2lca_helper
```
Comment (bmesuere) [inline app/controllers/api/api_controller.rb]: move to private section

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#discussion_r16986) by @bmesuere on Thu Oct 11 2018 at 10:13._

File: config/routes.rb [hunk 2]
```diff
@@ -69,6 +69,10 @@
     match 'pept2lca' => "api#pept2lca", via: [:get, :post]
     match 'taxa2lca' => 'api#taxa2lca', via: [:get, :post]
     match 'pept2prot' => 'api#pept2prot', via: [:get, :post]
+	  match 'pept2funct' => 'api#pept2funct', via: [:get, :post]
```
Comment (bmesuere) [inline config/routes.rb]: tab/space

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#discussion_r16987) by @bmesuere on Thu Oct 11 2018 at 10:20._

**Comment by bmesuere** [comment]
Deployed on morty. A small remark, on https://morty.ugent.be/apidocs the new endpoints need to be added

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80720) by @bmesuere on Thu Oct 18 2018 at 14:46._

**Comment by bmesuere** [comment]
@pverscha there seem to be a few errors:
```
An ActionView::Template::Error occurred in api#pept2ec:

  undefined method `[]' for nil:NilClass
  app/views/api/api/pept2ec.json.jbuilder:4:in `block in _app_views_api_api_pept_ec_json_jbuilder__4084974500524691144_70132864415420'


-------------------------------
Request:
-------------------------------

  * URL        : https://morty.ugent.be/api/v1/pept2ec.json?input[]=AAAAA,AEEAAGVAIVK
  * HTTP Method: GET
  * IP address : 128.101.135.251
  * Parameters : {"input"=>["AAAAA,AEEAAGVAIVK"], "controller"=>"api/api", "action"=>"pept2ec", "format"=>"json"}
  * Timestamp  : 2018-10-18 18:48:25 UTC
  * Server : morty
  * Rails root : /home/unipept/rails/releases/20181018123956
  * Process: 26827
```

```
A NoMethodError occurred in api#pept2funct:

  undefined method `[]' for nil:NilClass
  app/controllers/api/api_controller.rb:110:in `block in pept2funct'


-------------------------------
Request:
-------------------------------

  * URL        : https://morty.ugent.be/api/v1/pept2funct.json?input[]=AIPQLEVARPADAYETAEAYR&input[]=AAEVANADAIAAETTR
  * HTTP Method: GET
  * IP address : 128.101.135.251
  * Parameters : {"input"=>["AIPQLEVARPADAYETAEAYR", "AAEVANADAIAAETTR"], "controller"=>"api/api", "action"=>"pept2funct", "format"=>"json"}
  * Timestamp  : 2018-10-18 18:56:13 UTC
  * Server : morty
  * Rails root : /home/unipept/rails/releases/20181018123956
  * Process: 26827
```

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80748) by @bmesuere on Thu Oct 18 2018 at 21:30._

**Comment by pverscha** [comment]
Thanks! I'll try to fix them as soon as possible.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80750) by @pverscha on Thu Oct 18 2018 at 22:03._

**Comment by bmesuere** [comment]
```
An ActionView::Template::Error occurred in api#pept2lca:

  undefined method `id' for nil:NilClass
  app/views/api/api/_pept2lca.json.jbuilder:1:in `_app_views_api_api__pept_lca_json_jbuilder___600379135471140884_70132563452480'


-------------------------------
Request:
-------------------------------

  * URL        : https://morty.ugent.be/api/v1/pept2lca
  * HTTP Method: POST
  * IP address : 128.101.135.251
  * Parameters : {"equate_il"=>"true", "extra"=>"true", "names"=>"true", "input"=>["AVAAAVLGDSDAIQFVPTTGK", "ALEHGNVLLTGDEK", "DAFEAAGVDPDTDLSTWQNVDK", "DHVISVSNAPGLK", "ADYDLAVQK", "DIYPTLETAEGQDQALAK", "FTALQSGEIDMLS", "GLADAFATAYR", "AAEVANADAIAAETTR", "AASTASLEGTPGK", "GQLSNQAAVQR", "GFGFITEEGNNK", "EASTAELQEK", "DIPNVGEEALR", "IDGITEDIMSIALK", "EIANTASSDFYDGLENMK", "EEAIK", "DPAFFEPLSVATDK", "AVEVGSWQR", "GEFSANNQAPGK", "FLDLDPNNIESINVLK", "HVVDAGLAAGDGSNVGSVLYNR", "FAVAGGGGENAMTVLK", "GNFGYNAASEEYGDMLK", "CDIITSEPLINYR", "AGAWYSVEGER", "GISYEPVAFEK", "GYTSPYFVTNADK", "APGIQCAAGELATCLK", "DNLSIAINGGHISK", "GFGFITPAEGGNDLFVHYSEIK", "AGDQIISGADAPIK", "DGSYIINPTYSDLK", "EEVESEVNK", "HGDDQWHDIGK", "IETYTDPTTGASIEAQPNR", "ANGGLVTGADGIER", "AGANDPSEHVV", "IEDTSSDYDR", "AEEAAGVAIVK", "AGSVDLDEALFGLEPR", "GFAINEVIQLPK", "GEVVAVGPGK", "CNNCGDESELSR", "ANVEAFEAR", "FTMLNSGEGDILFR", "ELGYDMVGGAYR", "GQLPTNAGSSVVDDK", "DSPETDLTR", "FATENPDLLR", "ALQQSCLQAWK"], "controller"=>"api/api", "action"=>"pept2lca"}
```

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80751) by @bmesuere on Thu Oct 18 2018 at 22:49._

**Comment by pverscha** [comment]
Sequences that are not found in the database, are not part of the ``@result`-variable` that holds all data that should be outputted by these API-endpoints. Because we iterate over ``@input_order`` (which does hold all requested sequences, also the ones that aren't present in the db) and query ``@result``, we get a `NilClass` exception.

My question is now if we should just omit the variables not found from the API-answer, or whether we should include them but set `total_protein_count` to 0 and return both an empty `ec` and `go` list for these peptides?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80754) by @pverscha on Fri Oct 19 2018 at 08:05._

**Comment by bmesuere** [comment]
I think we don't return those sequences for the other calls, so I wouldn't return them here.
https://api.unipept.ugent.be/api/v1/pept2lca.json?input[]=DFDFDFDDF vs https://morty.ugent.be/api/v1/pept2lca.json?input[]=DFDFDFDDF

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80756) by @bmesuere on Fri Oct 19 2018 at 08:56._

**Comment by pverscha** [comment]
Alright! I'll fix it right away :)

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80760) by @pverscha on Fri Oct 19 2018 at 09:25._

**Comment by pverscha** [comment]
The issues mentioned above should all be fixed now :)

_[Original comment](https://github.ugent.be/unipept/unipept/pull/658#issuecomment-80816) by @pverscha on Sat Oct 20 2018 at 13:49._

**Pull request merged**